### 一. 基础知识练习
___

#### 1. 函数的定义和调用

In [3]:
# 定义一个名称为 draw_rect 的函数，功能是绘制指定大小（高度和宽度）的星号矩形，高度和宽度是绘制矩形的参数
def draw_react(height,width):
    for i in range(height):
        print("*" * width)

# 调用 draw_rect 函数，绘制高度为 5，宽度为 10 的星号矩形
draw_react(5,10) 

**********
**********
**********
**********
**********


#### 2. 参数的传递

##### 2.1 位置参数、关键字参数

In [10]:
def draw_rect(width,height,star):
    for i in range(height):
        print(star*width)

# 位置参数
draw_rect(5,3,"*")
print()

# 关键字参数
draw_rect(width=5,height=3,star="*")
print()

# 位置参数和关键字参数混合使用
draw_rect(5,height=3,star="*")

# 错误示范
# draw_rect(width=5,3,"*")

*****
*****
*****

*****
*****
*****

*****
*****
*****


#### 2.2 可选参数（默认值参数）

In [12]:
def draw_react3(width,height,star = '*'):
    for i in range(height):
        print(star * width)

# 不写第三个实参
draw_react3(5,6)
print()
# 写第三个实参
draw_react3(5,6,'#')

*****
*****
*****
*****
*****
*****

#####
#####
#####
#####
#####
#####


##### 2.3 变长参数

In [16]:
def order_dishes(customer,*dishes,num_people,**addtional_comments):
    print(f"欢迎您！{customer}")
    print(f"您预定的菜单为：{dishes}")
    print(f"进餐人数是{num_people}人")
    print(f"您附加的要求为{addtional_comments}")

# 模拟订餐
order_dishes("张先生","土豆炖牛腩","奥尔良烤翅","家常豆腐",num_people=3,口味="偏清淡",餐桌位置="偏僻角落")

# 说明：关键字参数的键必须是变量名，而不是字符串，但是输出后会自动加上引号

欢迎您！张先生
您预定的菜单为：('土豆炖牛腩', '奥尔良烤翅', '家常豆腐')
进餐人数是3人
您附加的要求为{'口味': '偏清淡', '餐桌位置': '偏僻角落'}


##### 2.4 解包参数


In [18]:
# 利用 * 将列表或元组解包为多个参数
def func1(a,b,c):
    print(a,b,c)
    return a + b + c

args = [1,2,3]
res = func1(*args)
print(res)

1 2 3
6


In [19]:
# 利用 ** 将字典键值对解包为多个关键字参数
def func2(x,y,z):
    print(x,y,z)
    return x + y + z

kargs = {"x":1,"y":2,"z":3}
res = func2(**kargs)
print(res)

1 2 3
6


##### 2.6 传递列表

a. 一个对照试验


In [32]:
# 定义函数increment_count
def increment_count(count:int, increment:int):
    """使count的值增加increment"""
    print(f"执行增量前，count的值是{count}, 标识是{id(count)}")
    count += increment
    print(f"执行增量后，count的值是{count}, 标识是{id(count)}")

# 定义函数increment_counts
def increment_counts(counts:list, increment:int):
    """使counts中所有元素的值增加increment"""
    print(f"执行增量前，counts的值是{counts}, 标识是{id(counts)}")
    for i in range(len(counts)):
        counts[i] += increment
    print(f"执行增量后，counts的值是{counts}, 标识是{id(counts)}")

# 对比实验
count_value = 100
print(f"调用增量函数前，count_value的值是{count_value}, 标识是{id(count_value)}")
increment_count(count_value, 10) # 把整数100的引用传递给形参count
print(f"调用增量函数后，count_value的值是{count_value}, 标识是{id(count_value)}")
print()

count_values = [100, 200, 300]
print(f"调用增量函数前，count_values的值是{count_values}, 标识是{id(count_values)}")
increment_counts(count_values, 10) # 把列表对象[100, 200, 300]的引用传递给形参counts
print(f"调用增量函数后，count_values的值是{count_values}, 标识是{id(count_values)}")


调用增量函数前，count_value的值是100, 标识是140729707804184
执行增量前，count的值是100, 标识是140729707804184
执行增量后，count的值是110, 标识是140729707804504
调用增量函数后，count_value的值是100, 标识是140729707804184

调用增量函数前，count_values的值是[100, 200, 300], 标识是2318735827904
执行增量前，counts的值是[100, 200, 300], 标识是2318735827904
执行增量后，counts的值是[110, 210, 310], 标识是2318735827904
调用增量函数后，count_values的值是[110, 210, 310], 标识是2318735827904


说明：
- 整数 100 是不可变对象，函数操作对100没有影响，也不会改变 `count_value` 对100的引用，执行增量后，`count` 引用了一个新对象，其值为110，其标识不同于100的标识
- 列表 [100,200,300] 为可变对象，函数操作对其有影响，执行增量前后，`count_values` 和 `counts` 引用的都是这个列表对象

b. 随机混排给定列表的元素值

In [33]:
import random
def shuffle(lst):
    for i in range(len(lst)):
        j = random.randint(0,len(lst)-1)
        lst[i],lst[j] = lst[j],lst[i]

# 对shuffle函数进行测试
def test_shuffle():
    v = [1,2,3,4,5]
    shuffle(v)
    print(v)

test_shuffle()

[5, 1, 4, 3, 2]


c. 来访者登记簿程序：定义函数 `check_in` 登记来访者的姓名和电话，并把记录保存到一个来访者列表


In [34]:
def check_in(visitors:list):
    """记录来访者输入的姓名和联系电话，追加到visitors"""
    name = input("请输入姓名：")
    tel = input("联系电话：")
    visitors.append((name,tel))

visitor_book = []    # 定义列表用来保存即将到来的来访者信息

# 假设有 3 名来访者的信息要登记，'_'用作临时循环变量
for _ in range(3):
    check_in(visitor_book)

print(visitor_book)

[('沙', '123456'), ('夏', '234567'), ('灿', '345678')]


#### 3. 函数的返回值

##### 3.1 返回简单值
定义函数circle_area，根据半径计算圆的面积

In [22]:
# 定义函数circle_area，根据半径计算圆的面积
import math

def circle_area(r):
    return math.pi*r**2

radius = eval(input("请输入圆的半径为：r = "))
area = circle_area(radius)
print(f"半径为 {radius} 的圆的面积为 {area:.2f}")


半径为 4 的圆的面积为 50.27


##### 3.2 返回逻辑值
定义函数 is_even 检测任意一个大于零的整数是否为整数，返回检测结果，是一个布尔型的逻辑值

In [24]:
# 定义函数 is_even 检测任意一个大于零的整数是否为整数，返回检测结果，是一个布尔型的逻辑值
def is_even(n):
    if n > 0 and n % 2 == 0:
        return True
    else:
        return False

num = int(input("请输入一个大于零的偶数："))
if is_even(num):
    print(f"{num} 是大于 0 的偶数")
else:
    print("不满足条件：不是大于零的偶数")

4 是大于 0 的偶数


##### 3.3 包含多个 `return` 语句的函数
定义函数is_prime判断一个数是否为素数，编写测试代码判断并输出1~99中的素数

In [28]:
# 定义函数is_prime判断一个数是否为素数，编写测试代码判断并输出1~99中的素数
def is_prime(n):
    if n < 2:
        return False
    elif n == 2:
        return True
    else:
        i = 2
        while i <= n // 2 + 1:
            if n % i == 0:
                return False
            else:
                i += 1
        return True

prime_lst = []
for j in range(1,100):
    if is_prime(j):
        prime_lst.append(j)

print(f"1~99 内的素数有 {','.join(map(str,prime_lst))}")


1~99 内的素数有 2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53,59,61,67,71,73,79,83,89,97


##### 3.4 返回元组（包含多个值）
定义一个函数describe，计算一组数的 3 个基本描述性统计量（最大值、最小值和平均值），返回这三个统计量

In [31]:
# 定义一个函数describe，计算一组数的 3 个基本描述性统计量（最大值、最小值和平均值），返回这三个统计量
def describe(*args):
    maximum = max(args)
    minimum = min(args)
    medium = sum(args)/len(args)
    return maximum,minimum,medium

lst = eval(input("请输入一组数据："))
print(lst)

describe_tup = describe(*lst)

print(f"最大值为 {describe_tup[0]}")
print(f"最小值为 {describe_tup[1]}")
print(f"平均值为 {describe_tup[2]}")

(1, 2, 3, 4)
最大值为 4
最小值为 1
平均值为 2.5


### 二. 实践作业
___

### 三. 案例指导
___